### DISCUSSION TITLE
MCPWM output on ESP32

### DISCUSSION THREAD

**Discussion by mcbp223** [pschatzmann/arduino-audio-tools#1412] [Ideas]
Hi, I was looking for a 3-level PWM solution for filterless, and it seems that the MCPWM peripheral in ESP32 had the up-down timer necessary to synthesize the waveforms.

I have attached an implementation tested on ESP32 and ESP32-S3. The disadvantage is that it doesn't have DMA, so it needs an interrupt for every sample, and another issue, the interrupt must be in a cpp file due to a linker error when it's header only.

Some other modifications I've made in library files:

- CodecMP3Helix.h, line 124 changed `addNotifyAudioChange(bi)` to `AudioDecoder::addNotifyAudioChange(bi)` to avoid an infinite recursive call (recent bug).
- SnapClient.h line 111 moved the call to `setupMDNS()` to line 121, after the calls to `setServerIP()` and `setServerPort()` otherwise IP obtained by MDNS gets overridden; the issue appeared recently also.
- SnapProcessor.h line 221, added `server_ip` and `server_port` to `ESP_LOGE(...)`, to be able to notice discrepancies like the above.

[aat-mcpwm.zip](https://github.com/pschatzmann/arduino-audio-tools/files/14472611/aat-mcpwm.zip)

**Comment by pschatzmann** [comment]
I suggest that you submit a pull request for the Snap Client.
What is the advantage of your PWM implementation compared with [PWMAudioOutput](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_p_w_m_audio_output.html) ?

Thanks for informing me about the bug in CodecMP3Helix that I have introduce with the implementation of the addNotifyAudioChange method

**Reply by pschatzmann** [reply]
I just had a quick look at the setupMDNS().
You are right, the logic has not been updated - it should actually set the class variables...

**Reply by mcbp223** [reply]
> I suggest that you submit a pull request for the Snap Client.

It may take me some time to setup the repository, if it's faster for you to rectify the MDNS issue it would be preferable I think.


> What is the advantage of your PWM implementation compared with [PWMAudioOutput](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_p_w_m_audio_output.html) ?

I have tried it, on ESP32 it's based on the LEDC peripheral, but I wasn't able to get fluid audio with it, especially if connected to Snapclient; possibly it doesn't go up in frequency high enough. However the main characteristics I was looking for was 3-level PWM in bridge mode that is made possible with a up-down counter timer that the MCPWM peripheral has.

![class-D filter-free 1049071-fig6 filterless](https://github.com/pschatzmann/arduino-audio-tools/assets/24973142/0bad0541-c072-428a-b4e7-3ae1e57a1153)



> Thanks for informing me

Happy to contribute; your libraries are of great value to users, thank you as well.

**Reply by pschatzmann** [reply]
No worries, the snap client has been corrected already...

**Comment by pschatzmann** [comment]
Hmm, what is "3-level PWM in bridge mode" exactly and how would you use it ?

**Reply by mcbp223** [reply]
2-level is normal PWM, with one level at GND and one level at +Vcc. 3-level adds an additional level at -Vcc.

An explanation for example is in Texas Instruments' note [SLAA701A - LC Filter Design](https://www.ti.com/lit/an/slaa701a/slaa701a.pdf), section 2.1 explains traditional (AD) PWM modulation, and section 2.2 summarizes this 3-level PWM (BD modulation), that helps reducing EMI and pretty much allows eliminating LC filters on output.
